### PR TITLE
Dw 1986

### PR DESCRIPTION
**Pull request by dou9g**
Added additional sh script in order to install and start the Kafka / Zookeeper Broker service

### CODE REVIEW

File: .idea/misc.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (dan-hill2802) [reply .idea/misc.xml]: @dou9g I'll get you set up with a global `.gitignore` for JetBrains

File: generic_packer_template.json.j2 [hunk 2]
```diff
@@ -31,9 +31,30 @@
     "provisioners": [
       {% for provisioner in event['provisioners'] %}
         {
-          "type": "{{ provisioner['type'] or 'shell' }}",
-          "script": "{{ provisioner['script'] }}"
+                  "type": "shell",
```
Comment (mattburgess) [inline generic_packer_template.json.j2]: You've now taken a completely generic mechanism and made it be kafka-ami-building specific.  You need to get the thing that generates your Lambda payload file to populate the provisioner portion of the event JSON correctly.  The trick here is that https://github.com/dwp/ami-builder/blob/master/ami_builder.py#L50-L60 copies all of the scripts you put in the config S3 bucket to /tmp inside the Lambda.  You should then be able to put a bunch of file-provisioner configs in the event payload to have them copied from /tmp to wherever they need to be on the EC2 instance that packer is using to build the AMI from.  A bunch of script-provisioner configs in the event payload can then set permissions, etc and I think you'll be good to go.

File: kafka-broker/kafka-broker-install.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
+
+## Script to install a kafka & zookeeper broker
+
+# Install kafka & zookeeper
+	sudo useradd kafka -m
```
Comment (mattburgess) [inline kafka-broker/kafka-broker-install.sh]: There's no need for any of these lines to be indented

File: kafka-broker/rc.local [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (mattburgess) [inline kafka-broker/rc.local]: shell-based startup scripts are pretty much a thing of the past.  Could you try replacing this with a couple of systemd unit files instead please (one for zookeeper and another one for kafka-server which should depend on the zookeeper one).

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+# .tfstate files
+*.tfstate
+*.tfstate.*
+
+# .tfvars files
+*.tfvars
+
+
+# JetBrains
```
Comment (dan-hill2802) [inline .gitignore]: after setting up the global .gitignore you can remove the JetBrains entries from this file

File: kafka-broker/rc.local [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
```
[the remark this answers is not in the document]
Comment (marco-chessa) [reply kafka-broker/rc.local]: This has been done , zookeeper and kafka respond to systemctl (i.e. systemd) style commands.

File: .gitignore (hunk 5 above)
Comment (marco-chessa) [reply .gitignore]: done on Marco's IDE, file fixed

File: kafka-broker/rc.local (hunk 4 above)
Comment (marco-chessa) [reply kafka-broker/rc.local]: rc.local has been removed, a systemd hybrid approach is in place on this first iteration.

File: kafka-broker/kafka-broker-install.sh (hunk 3 above)
Comment (marco-chessa) [reply kafka-broker/kafka-broker-install.sh]: fixed as requested

File: generic_packer_template.json.j2 (hunk 2 above)
Comment (marco-chessa) [reply generic_packer_template.json.j2]: The provisioners section does require more work to become generic, for instance the current format will accept only type "shell", however it is necessary at least another type , that is "file" with different keys (i.e. not "script" for instance).
This should be addressed in another iteration.

File: kafka-broker-install.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	cd /usr/local
+	mkdir kafka && cd kafka
+	chown kafka:kafka /usr/local/kafka
+ 	curl "http://www-eu.apache.org/dist/kafka/1.0.2/kafka_2.12-1.0.2.tgz" -o /tmp/kafka.tgz
+	tar -xvzf /tmp/kafka.tgz --strip 1
+	cd /usr/local/kafka
+	chown -R kafka:kafka *
+
+ # Build Kafka config file
+       >/usr/local/kafka/config/server.properties
```
[the remark this answers is not in the document]
Comment (marco-chessa) [reply kafka-broker-install.sh]: This has been done by Ian.

File: kafka-broker-install.sh [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ 	echo "log.retention.check.interval.ms=300000"     >> /usr/local/kafka/config/server.properties
+ 	echo "zookeeper.connect=localhost:2181"           >> /usr/local/kafka/config/server.properties
+ 	echo "zookeeper.connection.timeout.ms=6000"       >> /usr/local/kafka/config/server.properties
+ 	echo "group.initial.rebalance.delay.ms=0"         >> /usr/local/kafka/config/server.properties
+        chown kafka:kafka                                    /usr/local/kafka/config/server.properties
+
+# Start Zookeeper & Kafka service
+	chmod u+x /usr/local/kafka/bin/zookeeper-server-start.sh
+	chmod u+x /usr/local/kafka/bin/kafka-server-start.sh
+	/usr/local/kafka/bin/zookeeper-server-start.sh /usr/local/kafka/config/zookeeper.properties &
```
[the remark this answers is not in the document]
Comment (marco-chessa) [reply kafka-broker-install.sh]: This has been corrected by Ian.

File: kafka-broker-install.sh [hunk 9]
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (marco-chessa) [reply kafka-broker-install.sh]: This has been done by Ian.

File: .gitignore [hunk 10]
```diff
@@ -1,3 +1,12 @@
...
+
+# .tfstate files
+*.tfstate
+*.tfstate.*
+
+# .tfvars files
+*.tfvars
+
+# CI
+aviator_pipeline.yml
```
Comment (dan-hill2802) [inline .gitignore]: add new line

File: kafka-broker/generic_packer_template.json.j2 [hunk 11]
```diff
@@ -27,14 +27,19 @@
...
+      {% for provisioner in event['provision_script_keys'] %}
+        {
+          "type": "shell",
+          "script": "{{download_dir}}/{{ provisioner }}"
+        }{{ "," if not loop.last }}
       {% endfor %}
     ]
     {% endif %}
-}
+}
```
Comment (dan-hill2802) [inline kafka-broker/generic_packer_template.json.j2]: add new line

File: kafka-broker/kafka-broker-install.sh [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+sudo chmod u+x             /usr/local/kafka/bin/zookeeper-server-start.sh
+sudo chmod u+x             /usr/local/kafka/bin/kafka-server-start.sh
+sudo chown kafka:kafka -R  /usr/local/kafka
+sudo chown kafka:kafka -R  /var/log/kafka
+sudo chown kafka:kafka -R  /tmp/zookeeper
+sudo chown kafka:kafka -R  /tmp/kafka-logs
+sudo chkconfig --add zookeeper
+sudo chkconfig --add kafka
+sudo systemctl start zookeeper
+sudo systemctl start kafka
```
Comment (dan-hill2802) [inline kafka-broker/kafka-broker-install.sh]: add new line

File: kafka-broker/provisioners/kafka-broker-provisioner [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (dan-hill2802) [inline kafka-broker/provisioners/kafka-broker-provisioner]: I don't think this file is needed anymore

Comment (marco-chessa) [reply kafka-broker/provisioners/kafka-broker-provisioner]: Agreed, re-added in haste when re-adding all that was removed during the weekend.

File: .gitignore (hunk 10 above)
Comment (marco-chessa) [reply .gitignore]: done as requested

File: kafka-broker/generic_packer_template.json.j2 (hunk 11 above)
Comment (marco-chessa) [reply kafka-broker/generic_packer_template.json.j2]: done as requested

File: kafka-broker/kafka-broker-install.sh (hunk 12 above)
Comment (marco-chessa) [reply kafka-broker/kafka-broker-install.sh]: done as requested

File: kafka-broker/provisioners/kafka-broker-provisioner (hunk 13 above)
Comment (marco-chessa) [reply kafka-broker/provisioners/kafka-broker-provisioner]: done as requested

**Pull request merged**